specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,228 @@
+            template-matching line matching.
+        model: `~astropy.modeling.Model`
+            The model to fit for the wavelength solution. Defaults to a linear model.
+        """
+        self._input_spectrum = input_spectrum
+        self._model = model
+        self._line_list = line_list
+        self.spectral_unit = spectral_unit
+        self._cached_properties = ['wcs',]
+        self.fitter = fitter
+        self._potential_wavelengths = None
+        self._catalog = catalog
+
+        # ToDo: Implement having line catalogs
+        self._available_catalogs = get_available_catalogs()
+
+        if isinstance(line_list, (list, np.ndarray)):
+            self._line_list = QTable([line_list], names=["pixel_center"])
+
+        if self._line_list["pixel_center"].unit is None:
+            self._line_list["pixel_center"].unit = u.pix
+
+        # Make sure our pixel locations are sorted
+        self._line_list.sort("pixel_center")
+
+        if (line_wavelengths is None and catalog is None
+                and "wavelength" not in self._line_list.columns):
+            raise ValueError("You must specify at least one of line_wavelengths, "
+                             "catalog, or 'wavelength' column in line_list.")
+
+        # Sanity checks on line_wavelengths value
+        if line_wavelengths is not None:
+            if "wavelength" in line_list:
+                raise ValueError("Cannot specify line_wavelengths separately if there is"
+                                 "a 'wavelength' column in line_list.")
+            if len(line_wavelengths) != len(line_list):
+                raise ValueError("If line_wavelengths is specified, it must have the same "
+                                 "length as line_pixels")
+            if not isinstance(line_wavelengths, (u.Quantity, QTable)):
+                raise ValueError("line_wavelengths must be specified as an astropy.units.Quantity"
+                                 "array or as an astropy.table.QTable")
+            if isinstance(line_wavelengths, u.Quantity):
+                line_wavelengths.value.sort()
+                self._line_list["wavelength"] = line_wavelengths
+            elif isinstance(line_wavelengths, QTable):
+                line_wavelengths.sort("wavelength")
+                self._line_list = hstack(self._line_list, line_wavelengths)
+
+        if catalog is not None:
+            if isinstance(catalog, list):
+                self._catalog = catalog
+            else:
+                self._catalog = [catalog]
+            for cat in catalog:
+                if isinstance(cat, str):
+                    if cat not in self._available_catalogs:
+                        raise ValueError(f"Line list '{cat}' is not an available catalog.")
+
+            # Get the potential lines from any specified catalogs to use in matching
+            self._potential_wavelengths = concatenate_catalogs(self._catalog)
```

Comment:
Trying to input anything for `catalog` would get caught by the `ValueError` above since there aren't any available yet, but I agree that it would be better as a `NotImplementedError` for now.